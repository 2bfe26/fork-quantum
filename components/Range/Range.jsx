import PropTypes from 'prop-types';
import React from 'react';

import { ErrorMessage, Label, FieldGroup } from '../shared';

class Range extends React.Component {
  constructor(props) {
    super(props);

    const { value } = props;

    this.state = { value };
  }

  componentWillUpdate(nextProps) {
    if (nextProps.value !== this.state.value) {
      this.state.value = nextProps.value;
    }
  }

  handleChange = (e) => {
    const { onChange } = this.props;
    const { target: { value } } = e;

    this.setState({ value: Number(value) });

    onChange(e, { value: Number(value) });
  }

  render() {
    const { id, label, error } = this.props;
    const { value } = this.state;

    return (
      <FieldGroup>
        { label && <Label htmlFor={id}> {label} </Label> }
        <input
          {...this.props}
          id={id}
          value={value}
          type="range"
          onChange={this.handleChange}
        />
        {error && <ErrorMessage>{error}</ErrorMessage>}
      </FieldGroup>
    );
  }
}

Range.defaultProps = {
  error: '',
  id: '',
  label: '',
  onBlur: () => {},
  onChange: () => {},
  onFocus: () => {},
  value: 0,
};

Range.propTypes = {
  /** Display an error message */
  error: PropTypes.string,
  /** HTML identification */
  id: PropTypes.string,
  /** Display a label text that describe the field */
  label: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  value: PropTypes.number,
};

export default Range;
