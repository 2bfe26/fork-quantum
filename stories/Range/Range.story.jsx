
import React from 'react';
import { storiesOf } from '@storybook/react';
import Heading from '../../.storybook/components/Heading';
import AutoPropsApi from '../../.storybook/components/AutoPropsApi';
import ComponentPanel from '../../.storybook/components/ComponentPanel';
import { TabbedView, Tab } from '../../.storybook/components/TabbedView';
import Range from '../../components/Range';
import Atom from '../static/atom.svg';

storiesOf('3. Forms', module)
  .add('Range', () => (
    <Heading image={Atom} title="Range">
      <TabbedView>
        <Tab title="Usage">
          <ComponentPanel component={<Range />} importModules="Range" />
        </Tab>

        <Tab title="API">
          <AutoPropsApi component={Range} />
        </Tab>
      </TabbedView>
    </Heading>
  ));
