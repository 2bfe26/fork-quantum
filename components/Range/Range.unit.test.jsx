import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Range from './Range';

describe('Input component ', () => {
  it('should match the snapshot', () => {
    expect(renderer.create(<Range />).toJSON()).toMatchSnapshot();
  });

  it('should have value set in the state', () => {
    const value = 50;
    const wrapper = shallow(<Range value={value} />);

    expect(wrapper.state('value')).toEqual(value);
  });

  it('with an error message should match the snapshot', () => {
    expect(renderer.create(<Range error="Error message" />).toJSON()).toMatchSnapshot();
  });

  describe('with an "onChange" callback set', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<Range onChange={mockFn} />);
    const range = wrapper.childAt(0);
    const value = 30;
    const valueOb = { value };

    it('should initialize with default value', () => {
      expect(wrapper.state('value')).toEqual(0);
    });

    it('should call the callback and set a new value', () => {
      range.simulate('change', { target: valueOb });

      expect(wrapper.state('value')).toEqual(value);
      expect(mockFn).toHaveBeenCalledTimes(1);
      expect(mockFn).toHaveBeenCalledWith(valueOb);
    });
  });

  describe('with a label', () => {
    const range = <Range label="Text label" id="range-with-label" />;

    it('should match the snapshot', () => {
      expect(renderer.create(range).toJSON()).toMatchSnapshot();
    });

    it('should match label "htmlFor" label param with "id" range param', () => {
      const wrapper = shallow(range);
      const label = wrapper.childAt(0);
      const styledInput = wrapper.childAt(1);

      expect(label.prop('htmlFor')).toEqual(styledInput.prop('id'));
    });
  });
});
