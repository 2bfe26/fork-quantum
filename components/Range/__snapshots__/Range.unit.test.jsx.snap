// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component  should match the snapshot 1`] = `
<div
  className="Range__Wrapper-yyti5j-2 dPsrRO"
>
  <div
    className="Range__Dot-yyti5j-0 QSUov"
    disabled={false}
  />
  <div
    className="Range__Stripe-yyti5j-1 jfTQeM"
    disabled={false}
  />
</div>
`;
